Document auto-closing keywords for PRs. (#855)
python/devguide: getting-started/git-boot-camp.rst

6. Press the ``Create pull request`` button.

> You should include the issue number in the title of the PR (``gh-NNNNN: ...``).

Reviewer: ```suggestion
You should include the issue number in the title of the PR,
in the format ``gh-NNNNN: <PR Title>``.
```

Be more specific regarding the format, since at least on the most of the rest of GitHub itself, the conventional way to refer to GitHub issues numbers is using `#` rather than `gh-`.

You might also use `special keywords
<https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword>`_

> to automatically close the linked issue once the PR is merged, however you

Reviewer: ```suggestion
to automatically close the linked issue once the PR is merged. However, you
```

Fix comma splice, add missing comma and this makes more sense as two separate sentences.

6. Press the ``Create pull request`` button.

You should include the issue number in the title of the PR,
in the format ``gh-NNNNN: <PR Title>``.

You might also use `special keywords
<https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword>`_
to automatically close the linked issue once the PR is merged. However, you
should avoid them if other PRs need to be merged before the issue can be

> closed (e.g. backports to other branches).

Reviewer 2: For a new contributor, this might raise more questions than it answers. Can we link to a reference somewhere else in the devguide that tells contributors which PRs are generally backported, and which are not?

Reviewer: To note, there's also other benefits to linking a PR beyond just auto-close—most notably, it adds an easy to find reference to the PR from the issue and the issue from the PR a number of handy places in the GitHub UI, which make navigation a fair bit quicker than manually/semi-automatically trying to figure out the corresponding PR (or issue).

Reviewer 3: Nearly all bug fixes, doc fixes, and IDLE changes are backported, and nearly all feature changes are not.  But even feature issues may require multiple PRs, so the above advice is still incomplete.  And there are PRs without an issue.  By linking autoclose to linking of issue to PR, GH has made issue-PR linking worse than useless.

Reviewer: Yeah, unfortunately the story on GitHub around relationships of multiple PRs for one issue has never been good, and unlike a number of other pain points that have improved substantially in recent years, this one doesn't seem to have. And unlike UI issues, there's not as much Refined GitHub and tools like it can do about it—it is a big reason that GitHub overlays like ZenHub gained such a following

Author: > unfortunately the story on GitHub around relationships of multiple PRs for one issue has never been good

I created this:
* https://github.com/python/bedevere/issues/501

Reviewer 4: hey folks, I've captured this feedback! keep it coming :)

Reviewer: Thanks @omerbensaadon ! We all really appreciate it!

6. Press the ``Create pull request`` button.

You should include the issue number in the title of the PR,
in the format ``gh-NNNNN: <PR Title>``.

Linking to Issues and Pull Requests
-----------------------------------

You can link to issues and pull requests using ``gh-NNNNN`` (this form is
preferred over ``#NNNNN``).  If the reference appears in a list, the link
will be expanded to show the status and title of the issue/PR.

When you create a PR that includes ``gh-NNNNN`` in the title, `bedevere`_
will automatically add a link to the issue in the first message.

In addition, pull requests support `special keywords`_ that can be used to
link to an issue and automatically close it when the PR is merged.
However, issues often require multiple PRs before they can be closed (e.g.

> backports to other branches), so this features is only useful when
> a single PR is enough to address and close the issue.

Reviewer 2: ```suggestion
backports to other branches), so this features is only useful if
you know for sure that a single PR is enough to address and close the issue.
```

You should include the issue number in the title of the PR,
in the format ``gh-NNNNN: <PR Title>``.

Linking to Issues and Pull Requests
-----------------------------------

You can link to issues and pull requests using ``gh-NNNNN`` (this form is

> preferred over ``#NNNNN``).  If the reference appears in a list, the link

Reviewer: For commits and PR titles sure, but why in the comment body? `#NNNN` is the de-facto standard form on GitHub, and is naturally extended to issues on other repos and orgs than just the current one with `org/repo#NNN`, and references to issues/PRs from elsewhere, whereas I know of no such equivalent for `gh-NNN`.

Reviewer 5: > but why in the comment body

Probably for another potential migration in the future.

Older BPO issues also used hashed issue ids and they already cannot be automatically distinguished from later introduced GitHub PRs. So there's no need to continue such a confusion.

Reviewer: Thanks, that makes sense I guess. Unfortunately, even as a triager I can't mark my own review comments as resolved (unless/until python/core-workflow#460 is addressed), yet I _can_ mark others' comments as resolved as a PR author (which doesn't make any sense; if anything it should be the opposite), so feel free to resolve this one.

> will be expanded to show the status and title of the issue/PR.

Reviewer: Is that really what triggers this? I've gotten inconsistent results in my resent testing, with links sometimes expanding and sometimes not in and outside of lists, even in the same spot. I'd be very curious to know.

Also, `gh` (and `#`?) references in commit messages get expanded to `pythongh`—how does that happen?

Author: > Oh, is that really what triggers this?

I think so, and it took me a while to figure that out too.  ISTM that references like `#855` are consistently expanded in lists like:
* #855 
* with some text #855
1. #855
2. with some text #855
- [ ] #855
- [ ] with some text #855

But not in paragraphs like this (#855) or quotes like

> #855

Do you have any counterexamples?

> Also, gh (and #?) references in commit messages get expanded to pythongh—how does that happen?

I've seen that and I think it's just a bug in GitHub.

Reviewer: > with some text https://github.com/python/devguide/pull/855

Ah, I misinterpreted what you said originally to mean just bullets with no other text, when I was seeing them both with and without—silly me.

> Do you have any counterexamples?

Not any others that I can reliably reproduce—I believe the behavior I saw (on the What's New issues) when I was trying to figure this out might have perhaps just been a combination of an issue where the link wasn't getting created as above when the linked issue hadn't been created when I loaded the page, and/or just not being totally sure what caused it when testing a number of things (particularly `#NNN` vs `org/repo#NNN` vs `<full URL>` vs. <comment URL>).

Reviewer: One side note that came up during some related testing—AFAIK, `GH-`/`gh-` doesn't trigger the issue number lookup and autocompletion, which I find invaluable especially on repos nearing 6 digit issue and PR numbers like CPython.

Author: In https://github.com/python/devguide/pull/946#pullrequestreview-1096790696 the ref is not expanded, maybe because it's in a review?

Screenshot showing both the unexpanded list, and the popup on the first ref which confirms the ref is correct:
![image](https://user-images.githubusercontent.com/25624924/194470794-9b80c29a-85f2-42d4-8ca9-62d2466957dc.png)
